File: models/experimental/deit/tt/deit_encoder.py
# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

from typing import Optional, Tuple, Union

import torch
from torch import nn

import ttnn
from models.experimental.deit.tt.deit_config import DeiTConfig
from models.experimental.deit.tt.deit_layer import TtDeiTLayer


class TtDeiTEncoder(nn.Module):
    def __init__(self, config: DeiTConfig(), device, state_dict=None, base_address="") -> None:
        super().__init__()
        self.config = config
        self.layer = nn.ModuleList(
            [
                TtDeiTLayer(config, device, state_dict, f"{base_address}.layer.{_}")
                for _ in range(config.num_hidden_layers)
            ]
        )

        self.gradient_checkpointing = False

    def forward(
        self,
        hidden_states: ttnn.Tensor,
        head_mask: Optional[ttnn.Tensor] = None,
        output_attentions: bool = False,
        output_hidden_states: bool = False,
        return_dict: bool = True,
    ) -> Union[tuple]:
        all_hidden_states = () if output_hidden_states else None
        all_self_attentions = () if output_attentions else None

        for i, layer_module in enumerate(self.layer):
            if output_hidden_states:
                all_hidden_states = all_hidden_states + (hidden_states,)

            layer_head_mask = head_mask[i] if head_mask is not None else None

            if self.gradient_checkpointing and self.training:
                assert False, "No support for training yet!"
            else:
                layer_outputs = layer_module(hidden_states, layer_head_mask, output_attentions)

            hidden_states = layer_outputs[0]

            if output_attentions:
                all_self_attentions = all_self_attentions + (layer_outputs[1],)

        if output_hidden_states:
            all_hidden_states = all_hidden_states + (hidden_states,)

        output = tuple(v for v in [hidden_states, all_hidden_states, all_self_attentions] if v is not None)

        return output

File: models/experimental/deit/deit_cpp/tt_cpp/deit_encoder.cpp
// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
//
// SPDX-License-Identifier: Apache-2.0


#include "deit_encoder.h"
#include <stdexcept>

TtDeiTEncoder::TtDeiTEncoder(
    const DeiTConfig& config,
    std::shared_ptr<ttnn::MeshDevice> device,
    const std::unordered_map<std::string, torch::Tensor>& state_dict,
    const std::string& base_address
) : config(config), device(device), gradient_checkpointing(false) {
    
    // Initialize all encoder layers
    layers.reserve(config.num_hidden_layers);
    
    for (int i = 0; i < config.num_hidden_layers; ++i) {
        std::string layer_address = base_address + ".layer." + std::to_string(i) + ".";
        layers.push_back(
            std::make_unique<TtDeiTLayer>(config, device, state_dict, layer_address)
        );
    }
}

std::tuple<ttnn::Tensor, std::optional<std::vector<ttnn::Tensor>>, std::optional<std::vector<ttnn::Tensor>>> 
TtDeiTEncoder::forward(
    const ttnn::Tensor& hidden_states,
    const std::optional<std::vector<ttnn::Tensor>>& head_mask,
    bool output_attentions,
    bool output_hidden_states,
    bool return_dict
) {
    // Initialize output containers
    std::optional<std::vector<ttnn::Tensor>> all_hidden_states = std::nullopt;
    std::optional<std::vector<ttnn::Tensor>> all_self_attentions = std::nullopt;
    
    if (output_hidden_states) {
        all_hidden_states = std::vector<ttnn::Tensor>();
    }
    
    if (output_attentions) {
        all_self_attentions = std::vector<ttnn::Tensor>();
    }
    
    // Current hidden states
    ttnn::Tensor current_hidden_states = hidden_states;
    
    // Process through all layers
    for (size_t i = 0; i < layers.size(); ++i) {
        // Store hidden states if requested
        if (output_hidden_states) {
            all_hidden_states->push_back(current_hidden_states);
        }
        
        // Get layer head mask if provided
        std::optional<ttnn::Tensor> layer_head_mask = std::nullopt;
        if (head_mask.has_value() && i < head_mask->size()) {
            layer_head_mask = (*head_mask)[i];
        }
        
        // Forward pass through current layer
        if (gradient_checkpointing && false) { // Training not supported yet
            throw std::runtime_error("No support for training yet!");
        } else {
            auto layer_outputs = layers[i]->forward(
                current_hidden_states,
                layer_head_mask,
                output_attentions
            );
            
            // Extract outputs
            current_hidden_states = std::get<0>(layer_outputs);
            auto attention_weights = std::get<1>(layer_outputs);
            
            // Store attention weights if requested
            if (output_attentions && attention_weights.has_value()) {
                all_self_attentions->push_back(*attention_weights);
            }
        }
    }
    
    // Store final hidden states if requested
    if (output_hidden_states) {
        all_hidden_states->push_back(current_hidden_states);
    }
    
    return std::make_tuple(
        current_hidden_states,
        all_hidden_states,
        all_self_attentions
    );
}